### ISSUE TITLE
build progress not updating in terminal 

### ISSUE THREAD

**Issue by s-kris** [Elderjs/elderjs#152]
As you can see in the screenshot below, build is completed successfully but the progress bars are not updated. 
I want to know if I'm doing anything wrong.

I'm fetching data from cms in bootstrap hook and adding it to data object as documented in "Fetching External Data" section. In each route.js, I'm reading the externalData object from {data} to build the page. 

Thank you!

![Screenshot 2021-03-17 at 4 19 02 PM](https://user-images.githubusercontent.com/3975305/111459709-103d2a00-8741-11eb-9a03-65fcaf38aef8.png)

**Comment by nickreese** [comment]
@s-kris I believe you’re seeing a race condition. 

The issue is that it spins up child processes across all cores and only starts updating the bars after short delay once that is done. 

You’re building faster than that total delay is my bet.

**Comment by s-kris** [comment]
Oh my :)

**Comment by s-kris** [comment]
Also, can I know why the bootstrap hook is being run multiple times? The Idea is to fetch data once in that hook and re-use it.

Btw, I just quickly tried to fetch external data and set it on global object before starting the build to read it in bootstrap hook instead of network call. 
Even that function is called multiple times.

```
const { build } = require('@elderjs/elderjs');
const { preflightCheck } = require('./utils/preflight');

(async () => {
  await preflightCheck(); // fetch external data via network call and set it on global object here. 
  await build();
})();
```

**Comment by nickreese** [comment]
@s-kris Not sure 100% why your preflight would be called multiple times but `bootstrap` is called once for each worker that is spun up.

**Comment by s-kris** [comment]
I will check about the preflight, thanks!

So, I think it's really helpful to have a hook `beforBuildStart`, so that we can fetch external data here and re-use it in bootstrap instead of multiple network calls.

**Comment by s-kris** [comment]
> I will check about the preflight, thanks!

@nickreese 
okay, something is definitely weird. 
1. I cloned the sample elderjs repo with `npx degit Elderjs/template elderjs-app`. 
2. I added a log statement. 
3. I built it with `yarn build`
3. It's being called multiple times!
4. There's that progress bar issue/race as well.

```
const { build } = require('@elderjs/elderjs');

(async () => {
  console.log("this should not appear twice");
  await build();
})();
```

![Screenshot 2021-03-17 at 7 51 31 PM](https://user-images.githubusercontent.com/3975305/111484056-67e88f00-875b-11eb-9543-44db23a5c0f0.png)

**Comment by nickreese** [comment]
@s-kris Set your workers to 1 by adding this to your `elder.config.js` to see if you still see it twice.

```
build: {
    numberOfWorkers: 1, // This controls the number of worker processes spun up during build. It accepts negative numbers to represent the number of cpus minus the given number. Or the total number of processes to spin up. 
  }
```

My bet is that the `console.log` is getting called twice due to the cluster implementation... but I'd need to set aside more time to dig in to be sure. 

I'm open to a PR including a `beforeBuildStart` hook, but there be dragons here. The main issue is that serialization of data across processes is a bit complex and there are pitfalls. That said, the existing build process is pretty messy and could use a refactor if you're up for it.

**Comment by s-kris** [comment]
Sure @nickreese  Setting `numberOfWorkers` to 1 didn't fix it. 

I'm not very confident on handling the multi-process data-serialization, but I will anyways need to implement `beforeBuildStart` hack for my project. I'll go through elderjs code base and come back to you. I am already behind a PR (for stemap) :). Sorry, I'm a solo-founder. So..

**Comment by nickreese** [comment]
@s-kris As a prior solo founder, I'd just tell you to do any non-parallelizable work before you invoke Elder.js in your CI/CD process. Even if that is writing all of the API data to a json file that you toss out later. I went down the route of trying to implement a `beforeBuildStart` and after hitting edge-case after edge-case when it came to data parallelization... ultimately I just decided to preprocess things. 👍
